prov/rxm/src/rxm_cq.c

```diff
@@ -225,6 +169,116 @@ static int rxm_match_rma_iov(struct rxm_recv_entry *recv_entry,
 	return FI_SUCCESS;
 }
 
+static int rxm_finish_recv(struct rxm_rx_buf *rx_buf)
+{
+	int ret;
+
+	if (rx_buf->recv_entry->flags & FI_COMPLETION) {
+		FI_DBG(&rxm_prov, FI_LOG_CQ, "writing recv completion\n");
+		ret = ofi_cq_write(rx_buf->ep->util_ep.rx_cq,
+				   rx_buf->recv_entry->context,
+				   rx_buf->recv_entry->comp_flags |
+				   (rx_buf->recv_entry->flags & FI_MULTI_RECV) |
+				   ((rx_buf->pkt.hdr.flags & OFI_REMOTE_CQ_DATA) ?
+				    FI_REMOTE_CQ_DATA : 0),
+				   rx_buf->pkt.hdr.size,
+				   rx_buf->recv_entry->rxm_iov.iov[0].iov_base,
+				   rx_buf->pkt.hdr.data, rx_buf->pkt.hdr.tag);
+		if (ret) {
+			FI_WARN(&rxm_prov, FI_LOG_CQ,
+				"Unable to write recv completion\n");
+			return ret;
+		}
+	}
+	if (rx_buf->recv_entry->flags & FI_MULTI_RECV) {
+		struct rxm_rx_buf *unexp_rx_buf;
+		struct rxm_iov rxm_iov;
+
+		rx_buf->recv_entry->total_len -= rx_buf->pkt.hdr.size;
+
+		if (rx_buf->recv_entry->total_len <= rx_buf->ep->min_multi_recv_size) {
+			/* Since buffer is elapsed, release recv_entry */
+			rxm_recv_entry_release(rx_buf->recv_queue, rx_buf->recv_entry);
+			goto done;
+		}
+
+		FI_DBG(&rxm_prov, FI_LOG_CQ,
+		       "Repost Multi-Recv entry: "
+		       "consumed len = %zu, remain len = %zu\n",
+		       rx_buf->pkt.hdr.size,
+		       rx_buf->recv_entry->total_len);
+
+		rxm_iov = rx_buf->recv_entry->rxm_iov;
+		ret = rxm_match_iov(rxm_iov.iov, rxm_iov.desc, rxm_iov.count,	/* prev iovecs */
+				    rx_buf->pkt.hdr.size,			/* offset */
+				    rx_buf->recv_entry->total_len,		/* match_len */
+				    &rx_buf->recv_entry->rxm_iov);		/* match_iov */
+		if (OFI_UNLIKELY(ret))
+			goto done;
+
+		fastlock_acquire(&rx_buf->ep->recv_queue.lock);
+		unexp_rx_buf = rxm_check_unexp_msg_list(&rx_buf->ep->recv_queue,
+							rx_buf->recv_entry->addr,
+							0, 0);
+		if (unexp_rx_buf) {
+			dlist_remove(&unexp_rx_buf->unexp_msg.entry);
+			unexp_rx_buf->recv_entry = rx_buf->recv_entry;
+			fastlock_release(&rx_buf->ep->recv_queue.lock);
+			ret = rxm_cq_handle_data(unexp_rx_buf);
+			if (OFI_UNLIKELY(ret)) {
+				FI_WARN(&rxm_prov, FI_LOG_CQ,
+					"Unable to handle recv data\n");
+				goto done;
+			}
+		}
+
+		RXM_DBG_ADDR_TAG(FI_LOG_EP_DATA, "Enqueuing recv",
+				 rx_buf->recv_entry->addr, rx_buf->recv_entry->tag);
+		dlist_insert_tail(&rx_buf->recv_entry->entry, &rx_buf->ep->recv_queue.recv_list);
+		fastlock_release(&rx_buf->ep->recv_queue.lock);
```

Comment:
The above code looks similar to the one in rxm_ep_recv_common. Can it be shared in a single function?